server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -1553,14 +1557,23 @@ public Searcher(
             QueryCachingPolicy queryCachingPolicy,
             Closeable onClose
         ) {
-            super(reader);
+            // pass in an executor so we can shortcut and avoid looping through all segments
+            super(reader, source.equals(CAN_MATCH_SEARCH_SOURCE) ? Runnable::run : null);
```

Comment:
there's many more scenarios where all we need is an index reader. This is quite hard to untangle due to the many acquireSearcher callers and different variants of the method, wrapping going on as well. I am tempted to add an acquireIndexReader method to Engine, but this requires significant cleanup to become effective and better than what we currently have.

Reply:
On the other hand, if there is significant overhead on creating a searcher, perhaps that's something to fix in Lucene rather than working around in ES. What triggered this change is suspicions around overhead caused by creating slices eagerly, mostly stuff that happens in the IndexSearcher constructor. We should figure out how to minimize that in Lucene?

Reply:
I agree and https://github.com/apache/lucene/pull/14343 should help quite a bit on this front. I wonder though, maybe the problem isn't so much with the slice cost but with the fact that we don't really reuse searcher instances much. We don't really need a new searcher (for e.g. can_match) across queries so long as we don't see new data do we?

Reply:
That is definitely a possible improvement, but what I am trying to do is figure out the cause of a regression.  We have not changed how many times we create a searcher, but we did add overhead to the creation of a searcher.